validator/src/main.rs

```diff
@@ -0,0 +1,216 @@
+use std::rc::Rc;
+use std::str::FromStr;
+
+use anchor_client::solana_client::pubsub_client::PubsubClient;
+use anchor_client::solana_client::rpc_config::{
+    RpcSendTransactionConfig, RpcTransactionLogsConfig,
+    RpcTransactionLogsFilter,
+};
+use anchor_client::solana_sdk::commitment_config::CommitmentConfig;
+use anchor_client::solana_sdk::ed25519_instruction::{
+    DATA_START, PUBKEY_SERIALIZED_SIZE, SIGNATURE_SERIALIZED_SIZE,
+};
+use anchor_client::solana_sdk::signer::keypair::read_keypair_file;
+use anchor_client::solana_sdk::signer::Signer;
+use anchor_client::solana_sdk::{self};
+use anchor_client::{Client, Cluster};
+use anchor_lang::solana_program::instruction::Instruction;
+use anchor_lang::solana_program::pubkey::Pubkey;
+use base64::Engine;
+use bytemuck::bytes_of;
+use lib::hash::CryptoHash;
+use solana_ibc::{accounts, instruction};
+
+fn main() {
+    setup_logging();
+
+    let rpc_url = std::env::var("RPC_URL")
+        .unwrap_or_else(|_| "http://127.0.0.1:8899".to_string());
+    let ws_url = std::env::var("WS_URL")
+        .unwrap_or_else(|_| "ws://127.0.0.1:8900".to_string());
+    let program_id: String = std::env::var("PROGRAM_ID").unwrap_or_else(|_| {
+        "9fd7GDygnAmHhXDVWgzsfR6kSRvwkxVnsY8SaSpSH4SX".to_string()
+    });
+    let genesis_hash = std::env::var("GENESIS_HASH").unwrap_or_else(|_| {
+        "o0OboGSWBU0eJmu3A8mraKscUSOm5LaCKRWLW4IAANw=".to_string()
+    });
```

Comment:
Parse it here into CryptoHash so you don’t have to parse it in the loop later.

Reply:
well then, the `GENESIS_HASH` from the env should also be in `CryptoHash` right?

Reply:
or do you mean outside the loop?

Reply:
Yes. You can CryptoHash::from_base64 here.